File: benchmarks/values/bool_sieve.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

int main(void) {
    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    int n = 1000000;
    char *is_prime = malloc(n + 1);
    memset(is_prime, 1, n + 1);
    is_prime[0] = 0;
    is_prime[1] = 0;

    int i = 2;
    while ((long long)i * i <= n) {
        if (is_prime[i]) {
            int j = i * i;
            while (j <= n) {
                is_prime[j] = 0;
                j += i;
            }
        }
        i++;
    }

    int count = 0;
    for (int k = 0; k <= n; k++) {
        if (is_prime[k]) count++;
    }

    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed = (t1.tv_sec - t0.tv_sec) + (t1.tv_nsec - t0.tv_nsec) / 1e9;
    printf("%d\n", count);
    printf("elapsed: %.3fs\n", elapsed);
    free(is_prime);
    return 0;
}

File: benchmarks/values/bool_sieve.rs
use std::time::Instant;

fn main() {
    let t0 = Instant::now();

    let n: usize = 1_000_000;
    let mut is_prime = vec![true; n + 1];
    is_prime[0] = false;
    is_prime[1] = false;

    let mut i: usize = 2;
    while i * i <= n {
        if is_prime[i] {
            let mut j = i * i;
            while j <= n {
                is_prime[j] = false;
                j += i;
            }
        }
        i += 1;
    }

    let mut count: i64 = 0;
    for k in 0..=n {
        if is_prime[k] {
            count += 1;
        }
    }

    let elapsed = t0.elapsed();
    println!("{}", count);
    println!("elapsed: {:.3}s", elapsed.as_secs_f64());
}
